### PR TITLE
Create the password propagation (WIP)

### PR DESCRIPTION
**Pull request by halilozanakgul**
DESCRIPTION: Add password propagation for ALTER ROLE and master_add_node().

### CODE REVIEW

File: src/backend/distributed/commands/role.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	passwordDatum = heap_getattr(tuple, Anum_pg_authid_rolpassword,
+								 pgAuthIdDescription, &isNull);
+
+	table_close(pgAuthId, AccessShareLock);
+	ReleaseSysCache(tuple);
+
+	initStringInfo(&setPasswordQueryBuffer);
+	appendStringInfo(&setPasswordQueryBuffer,
+					 "SELECT set_password_if_role_exists('%s', '%s')",
+					 rolename,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/role.c]: Instead of manually adding single quotes, we rely on Postgres' `quote_literal_cstr ` function, which can handle edge cases when the name already contains some escaping characters. To test the necessity, create a user name with special characters in it (we should have tests around it): 

```SQL
CREATE ROLE  "weird'role_name";
alter role "weird'role_name" PASSWORD 'test'; 
DETAIL:  on server onderkalaci@localhost:9700
ERROR:  syntax error at or near "role_name"
CONTEXT:  while executing command on localhost:9700
```

File: src/backend/distributed/commands/role.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+
+bool EnablePasswordPropagation = true;
+
+/*
+ * ProcessAlterRoleStmt actually creates the plan we need to execute for alter
+ * role statement.
+ */
+List *
+ProcessAlterRoleStmt(AlterRoleStmt *stmt, const char *queryString)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/role.c]: I think this name is too broad, the `ALTER ROLE` could have lots of options, we're only considering `password` piece. So, we should probably rename it to `ProcessAlterRoleSetPasswordStmt()`

File: src/backend/distributed/sql/udfs/set_password_if_role_exists/9.1-1.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+CREATE OR REPLACE FUNCTION set_password_if_role_exists(
+    role_name regrole,
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/set_password_if_role_exists/9.1-1.sql]: If we use `regrole`, that'd defeat our purpose for `skip if user not exists`, because Postgres would automatically error out even before our code is called:
```SQL
SELECT set_password_if_role_exists('r2', 'md55ed7902b70147c4a6fb4517eaa0f24b7');;
ERROR:  role "r2" does not exist
LINE 1: SELECT set_password_if_role_exists('r2', 'md55ed7902b70147c4...

```

File: src/backend/distributed/commands/role.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+{
+	ListCell *optionCell = NULL;
+	List *commands = NIL;
+	const char *setPasswordQuery = NULL;
+
+	if (!IsCoordinator())
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/commands/role.c]: I tried to make this function a little more generic. It is for all `ALTER ROLE` statements, instead of just `ALTER ROLE ... PASSWORD` statements. So if we ever want to also propagate, let's say, `ALTER ROLE ... CREATEROLE` statements we will just have to add one more `if` in the `foreach`. So I put the check in the the `if`.

File: src/backend/distributed/commands/role.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+	foreach(optionCell, stmt->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		if (EnablePasswordPropagation && strcasecmp(option->defname, "password") == 0)
+		{
+			setPasswordQuery = CreateSetPasswordQueryForWorkerNodes(
+				stmt->role->rolename);
+			commands = lappend(commands, (void *) setPasswordQuery);
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/commands/role.c]: Same as former comment. I thought we might want to support propagating statements such as `ALTER ROLE ... PASSWORD '...' CREATEROLE`.

File: src/backend/distributed/commands/role.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
+#include "postgres.h"
+
+#include "nodes/parsenodes.h"
+#include "utils/syscache.h"
+#include "access/table.h"
+#include "catalog/pg_authid.h"
+#include "utils/rel.h"
+#include "utils/builtins.h"
+#include "access/heapam.h"
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/commands/role.c]: Done 👍

File: src/backend/distributed/commands/role.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	foreach(optionCell, stmt->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		if (EnablePasswordPropagation && strcasecmp(option->defname, "password") == 0)
+		{
+			setPasswordQuery = CreateSetPasswordQueryForWorkerNodes(
+				stmt->role->rolename);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/role.c]: Indeed, `ALTER ROLE CURRENT_USER PASSWORD 'test' VALID UNTIL '2020-01-01';` this is crashing.

Note that postgres allows you to use `CURRENT_USER` or `SESSION_USER ` in the `ALTER ROLE` command. We should resolve those cases to accurate user names. SQL api is the following, please find the C apis in the code:

```SQL
SELECT SESSION_USER, CURRENT_USER;
┌──────────────┬──────────────┐
│ session_user │ current_user │
├──────────────┼──────────────┤
│ onderkalaci  │ onderkalaci  │
└──────────────┴──────────────┘
(1 row)
```

**Comment by onderkalaci** [comment]
superseded by #3184, so closing
